V8: Add the AngularJS directive template url pathname into the template

nathanwoulfe opened the pull request:
… view as part of the build

### Prerequisites

- [x] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes #5047 and fixes #6003 

### Description

Changes here add a step to the `views` gulp task in which the view being copied is modified to add a `data-filepath` attribute to the root node (or an HTML comment if the view happens to be empty, which some are).

Changes can be verified by inspecting the backoffice and searching for 'filepath', and doing the same in directives.js.

![filepath](https://user-images.githubusercontent.com/3248070/60643510-a4d0fd00-9e76-11e9-8e36-0bfc88da4d8a.png)

These changes don't introduce any functional changes, they just add the new attribute which will be really really helpful for backoffice devs. Changes do introduce a couple of new NPM packages to the build chain, being gulp-modify-file and jsdom, both as dev dependencies.

The way the current build is run, tasks are run in parallel where possible, but that needed to change as the `js` task is now dependent on the output of `views` - since the directive templates are inlined (#4871), in the `js` task, we need to reference the files output by `views`, from their final location.

Changing `views`  to run in series doesn't add any noticeable time to the build, but not entirely sure that referencing the files in a different project is the best approach - open to suggestions. Could also chop a bit more off the file path too.

Updates to include a solution for #6003 are explained [down further](https://github.com/umbraco/Umbraco-CMS/pull/5785#issuecomment-517539524)

emmaburstow commented:
Hiiiiiya Nathan!

Thanks for the work - we'll get a look at it and ask if we've any questions.

Em

nathanwoulfe commented:
Picking up on @PeteDuncanson's suggestion in #6003, I've updated this to clean out the physical template files when the template content has been inlined into the respective directives. Why? Avoids shipping a stack of html files that aren't doing anything - it's not a huge saving size-wise, but removes 110 files and 22 directories from /umbraco/views (on build).

How does it work? It extends on (and depends on) the previous changes I made to add the file path as an attribute in the template. With that in place, I've added a new 'clean' task to the Gulp config, which runs at the end of the build process. First, it grabs all the compile JS files, and extracts the value from any data-filepath attributes in the file content (which will only appear in an inline template string, which has been embedded as part of the JS processing tasks). Armed with a collection of path segments, I'm then deleting the files from the output folder, because if their template string exists in the JS, the physical file isn't needed.

Next, a separate task recursively deletes empty directories from /umbraco/views, since empty directories are empty.

The clean up tasks only run on `gulp build` - for `dev` and `fastdev` there's no benefit in removing the files, so left them in place. Also makes sense that dev environment would make more use of the data-filepath attribute, so having the files at that location is sensible.

👍
